### DISCUSSION TITLE
Forward/BackwardResource in OperatorProperty Interface

### DISCUSSION THREAD

**Discussion by ArmageddonKnight** [apache/incubator-mxnet#13312] [Q&A]
Hi,

Sorry but I have a question regarding the `OperatorProperty` interface. Let us take the `RNNProp` as an example:

```C++
class RNNProp : public OperatorProperty
{
  ...
  std::vector<ResourceRequest> ForwardResource(
      const std::vector<TShape> &in_shape) const override {
    return {ResourceRequest::kTempSpace};
  }

  std::vector<ResourceRequest> BackwardResource(
      const std::vector<TShape> &in_shape) const override {
    return {ResourceRequest::kTempSpace};
  }
  ...
```

I noticed we do not declare ahead of time the shape or the size of temporary workspace that we need. It was not until the actual forward/backward propagation when we say:

```C++
    Tensor<gpu, 1, DType> temp_space =
      ctx.requested[rnn_enum::kTempSpace].get_space_typed<gpu, 1, DType>(
                              mshadow::Shape1(temp_size), s);
```

I have two questions regarding the temporary workspace management:

- Am I right in saying that **the length of the vector returned by the `ForwardResource` and `BackwardResource` method is equal to the number of temporary workspace tensors I can use in the forward and backward propagation, respectively?**
- **Why don't the `ForwardResource` and `BackwardResource` return a list of _shapes_ that indicate the shapes of the workspace**? IMHO this information can be helpful in managing the workspace (allocation/deallocation/reuse) ahead of time.

**Comment by zachgk** [comment]
@mxnet-label-bot [Question, Operator, C++]

For questions, please submit them on MXNet discussion forum (https://discuss.mxnet.io), where it will get a wider audience and allow other to learn as well.
 I'm closing this issue now in favor of the discussion forum issue you will file, please feel free to re-open if closed in error.
 Thanks!

**Comment by zachgk** [comment]
@mxnet-label-bot add [Question, Operator, C++]

**Comment by leleamol** [comment]
Removing the C++ keyword as the question is related to custom operators and not the C++ API in cpp-package.

@mxnet-label-bot remove [C++]
